Replace the `operation` string in Conv2dGenerator with an enum.

krzysz00 opened the pull request:
This guards against "spoooky action at a distance" crashes from
unexpected values for the convolution operation.

krzysz00 committed:
Replace the `operation` string in Conv2dGenerator with an enum.

This guards against "spoooky action at a distance" crashes from
unexpected values for the convolution operation.

jerryyin commented on mlir/include/mlir/Dialect/MIOpen/Generator/Conv2dGenerator.h:
```diff
@@ -17,14 +17,20 @@
+
 class Conv2dGenerator {
 public:
+  enum Op { Fwd, Dummy, BwdData, BwdWeight };
```
Could you think of a way to consolidate this enum with another, use a separate header if necessary.

https://github.com/ROCmSoftwarePlatform/llvm-project-mlir/blob/2c09afa99dca7bcac3e2030cd84b80bf489af663/mlir/include/mlir/Dialect/MIOpen/MIOpenOps.h#L28

krzysz00 replied:
I didn't know we had that - I'll try and consolidate once @whchung gets back to us about the dummy op

whchung replied:
Let’s remove conv2d_dummy. That was introduced to showcase how we can split one convolution into multiple ones.

jerryyin commented on mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp:
```diff
@@ -50,11 +50,17 @@ static cl::opt<std::string> outputFilename("o", cl::desc("Output filename"),
+    "operation", cl::desc("Convolution operation,"),
+    cl::values(clEnumValN(Conv2dGenerator::Op::Fwd, "conv2d",
+                          "Forward convolution"),
+               clEnumValN(Conv2dGenerator::Op::Dummy, "conv2d_dummy",
+                          "Pretend to perform a forward convolution"),
```
I believe we shouldn't allow user to populate a conv2d dummy op

jerryyin replied:
Actually maybe conv2d dummy op shouldn't exist at all, though I'm not sure. @whchung: Are we still using the conv2d dummy op?

krzysz00 committed:
Remove Conv2DDummyOp

This operation isn't used anywhere or important for anything, and, per
Jack, is a leftover from old code.

Removing this operation allows us to use the ConvOpType enum that
already existis in the MIOpen dialect.

jerryyin commented on mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp:
```diff
@@ -118,6 +117,31 @@ Type Conv2dGenerator::getDataType(OpBuilder &builder) const {
   return dataType;
 }
 
+Optional<miopen::ConvOpType>
+Conv2dGenerator::getOpForName(const StringRef name) {
+  if (name == "conv2d") {
+    return miopen::Conv2DOpType;
+  }
+  if (name == "conv2d_bwd_data") {
+    return miopen::Conv2DBwdDataOpType;
+  }
+  if (name == "conv2d_bwd_weight") {
+    return miopen::Conv2DBwdWeightOpType;
+  }
+  return llvm::None;
+}
+
+const char *Conv2dGenerator::getNameForOp(const miopen::ConvOpType op) {
+  switch (op) {
+  case miopen::Conv2DOpType:
+    return "conv2d";
+  case miopen::Conv2DBwdDataOpType:
+    return "conv2d_bwd_data";
+  case miopen::Conv2DBwdWeightOpType:
+    return "conv2d_bwd_weight";
+  }
+}
+
```
Nit: I think it is a better idea for those two convenient functions to stay close to the `miopen::ConvOpType`. I believe we did this sort of translation in `LowerMIOpenOps.h` as well. But for the scope of this review, just add the convenience function to where the type is should be enough.

krzysz00 replied:
And fixed.

jerryyin commented on mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp:
```diff
@@ -118,6 +117,31 @@ Type Conv2dGenerator::getDataType(OpBuilder &builder) const {
+    return "conv2d_bwd_data";
+  case miopen::Conv2DBwdWeightOpType:
+    return "conv2d_bwd_weight";
+  }
```
Nit: Extendibility-wise, it is probably easier if you handle it the same way: use optional such that this reminds a developer certain situation needs to be handled when a new operation type is added, for example.

krzysz00 replied:
Well, unlike the string->enum function, this one is total, so there's no point in returning None - it'll never happen. And if someone adds a new case to the enum, they'll get compiler warnings about this one.

jerryyin commented:
The state as-is is looking pretty good. I don't worry much about the two nits and I'm ready to give approval/merge if you'd prefer the current state.

krzysz00 committed:
Move type conversion functions

jerryyin reviewed (approved):
LGTM, thanks!
